### ISSUE TITLE
lifetime problems with self-referential struct

### ISSUE THREAD

**Issue by LU15W1R7H** [ajrcarey/pdfium-render#44]
Hi, first off thanks for this library. It's really nice to be able to use such a good pdf renderer like pdfium in rust.
I'm integrating pdfium into my digital pen note taking application, which needs to re-render the pdf potentially every frame (when scaling, rotating or translating the page).

I have major issues with the API that is provided by pdfium-render. There are lifetimes everywhere (which in itself might be justified) but the most function signatures don't have explicit lifetimes and as a consequence all lifetimes are the same, which is overly restrictive. This is probably rather easy to solve.

Furthermore the `PdfDocument` requires a bytes slice when loading from bytes. This means the document can't own the bytes and I have to store them separately from the document. This results in a self referential struct that contains bot the document and the bytes.
It's not a solution to re-parse the document every time from the bytes as this is too slow. Do you have a solution for this problem?

I would be willing to do some contributions to this project as I heavily rely on it and need it for my own project.

Edit: After giving this all some more thought and taking a look at some functions in pdfium-render, I now think that these restrictions are from the C++ pdfium library. That's just how it is designed, right? It just happens to not align very well with the way the borrow checker would like you to write rust code.

**Comment by ajrcarey** [comment]
Hi Luis,

Good to hear from you. I'm very happy to help you with this.

Let's deal with your second problem first. `Pdfium::load_from_bytes()` takes a byte slice for two main reasons: first, it's the simplest way of taking data from the `include_bytes!()` macro (since that returns a slice, not a `Vec`), and second, it mirrors the memory management approach of Pdfium itself, where the caller retains responsibility for memory ownership of the PDF byte data, not Pdfium. (Pdfium only ever takes responsibility for the memory management of the data structures it creates itself.)

I don't see a problem with adding an alternate function that would take ownership of the PDF byte data and stash it somewhere inside `PdfDocument`. But I'm curious as to why you end up with a self-referential struct. Can you tell me more about that? I'm not seeing the reason why you need to do that off the top of my head.

**Comment by LU15W1R7H** [comment]
The reason for the self referential struct is the following.

I have the pdf data as a byte slice. In order to get a document I need to `Pdfium::load_pdf_from_bytes`, which takes a reference to the `Pdfium` struct (which uses the bindings).  Now in order to not re-parse the file every frame, I have to store it in a struct. This struct is self-referential.
```rust
pub struct PdfManager {
  pdfium: Pdfium,
  document: PdfDocument<'pdfium>,
}
```

**Comment by ajrcarey** [comment]
Ok, I must have misunderstood your original comment, which said you needed a self-referential struct to store both the PdfDocument and the bytes for that document. I didn't understand the reason for that and the example you provided doesn't reference the byte data at all, so I'm still not really following you. Why do you need to reload the byte data on every frame? Or, to look at it another way: why does `PdfDocument` need to take memory ownership of the byte data?

You are absolutely right that you need to keep your `Pdfium` object alive at least as long as any `PdfDocument` object created from that Pdfium object. Pdfium (and therefore your entire application) will segfault if you don't, so pdfium-render is actually helping you here by enforcing that lifetime requirement.

**Comment by LU15W1R7H** [comment]
> Ok, I must have misunderstood your original comment, which said you needed a self-referential struct to store both the PdfDocument and the bytes for that document. I didn't understand the reason for that and the example you provided doesn't reference the byte data at all, so I'm still not really following you. Why do you need to reload the byte data on every frame? Or, to look at it another way: why does PdfDocument need to take memory ownership of the byte data?

Sorry for confusing you. I kind of changed the example without saying anything. I managed to avoid the self-reference from `PdfDocument` to the bytes by using `Pdfium::load_pdf_from_file` instead of `Pdfium::load_pdf_from_bytes`. So I was able to avoid this problem.

But I didn't manage to avoid the self-reference I explained in the comment above.

**Comment by LU15W1R7H** [comment]
> You are absolutely right that you need to keep your `Pdfium` object alive at least as long as any `PdfDocument` object created from that Pdfium object. Pdfium (and therefore your entire application) will segfault if you don't, so pdfium-render is actually helping you here by enforcing that lifetime requirement.

Yes pdfium-render enforces the lifetime requirements from the C++ pdfium library correctly. But it makes `pdfium-render` for my uses unusable as I've explained above.

**Comment by ajrcarey** [comment]
Ok. So are you saying that the second problem you mentioned when opening the issue is no longer a problem?

The need to store these lifetime-bound objects across frames is the fundamental source of your problem, I think. Tell me more about your note-taking application. Is the UI written in immediate mode, or is it driven by an event loop? How do you store other long-lived state in your application across frames?

There may be justification here in dropping down to using the raw FPDF_* functions rather than pdfium-render's higher-level interface. I am happy to help you with that. I'm also happy to take specific feedback about individual functions with overly restrictive lifetime bounds, as you mentioned in your original comment.

**Comment by LU15W1R7H** [comment]
> Ok. So are you saying that the second problem you mentioned when opening the issue is no longer a problem?

This particular problem is solved for now but it's very similar to the other two cases of self-references.

**Comment by LU15W1R7H** [comment]
> There may be justification here in dropping down to using the raw FPDF_* functions rather than using pdfium-render's higher-level interface. I am happy to help you with that. I'm also happy to take specific feedback about individual functions with overly restrictive lifetime bounds, as you mentioned in your original comment.

> If you are absolutely, 100% certain that pages are not going to be mutated across frames, then perhaps your hashmap could store raw FPDF_PAGE handles rather than PdfPage high level objects. FPDF_PAGE handles, and other raw FPDF_* handles, are not subject to any formal lifetimes - they're just pointers - but you need to be very careful not to use them after they've been invalidated.

I really don't want to deal with the raw bindings because it introduces unsafety into my application.

**Comment by ajrcarey** [comment]
Thank you for the information on your project - it looks great.

On the face of it, introducing a lifetime such as

```
pub struct PdfManager<'a> {
    ...
    document: PdfDocument<'a>,
    pages: HashMap<usize, PdfPage<'a>>
}
```

could work. But that means your `PdfManager` struct now carries a lifetime, and I have the impression you _really_ want to avoid introducing lifetimes into your Application struct.

The fundamental issue here is that two separate components - your application, and Pdfium - need to share memory. There has to be agreement on lifetimes _somewhere_, it's just a question of where. This isn't anything specific to pdfium-render, it will happen anywhere two disparate components need to share memory. You will see similar approaches in many other Rust libraries dealing with FFI.

I've just seen your most recent comment where you say you don't want to use raw bindings because it's unsafe. I agree, it is! - because you are implicitly taking over responsibility for managing the lifetimes yourself, rather than having Rust do it for you. But it is possible to do this safely in simple situations (a single document that never mutates).

There is a really fundamental choice to be made here: do you let Rust manage the lifetimes for you, or do you do it yourself? I don't see a third option - please tell me if you do.

If it were me, I would introduce the lifetimes and carry them up through into the `Application` struct. It's more idiomatic, the intent is clearer, and the added complexity in practice is likely to be very minimal if there is only a single `Application` instance in use in your application; you may never need to worry about specifying the lifetime explicitly at all, as Rust is very good about inferring the lifetime in simple situations.

But if you absolutely _had_ to avoid the lifetimes, you could rework your `PdfManager` struct like this:

```
pub struct PdfManager {
    pdfium: Pdfium,
    document_handle: FPDF_DOCUMENT,
    pages: HashMap<usize, FPDF_PAGE>
}
```

and rework your document and page loading functions to use `FPDF_LoadDocument()` and `FPDF_LoadPage()` directly. Once you have those handles, it's easy enough to add functions to `PdfiumLibraryBindings` to let you turn those back into `PdfDocument` and `PdfPage` objects if you still want to use Pdfium's high-level interface - although, if you were just interested in rendering, personally I wouldn't bother, I'd just use `FPDF_RenderPage()` directly.

You would need to implement the `Drop` trait for your `PdfManager` struct to ensure the document and pages you loaded are closed, but this isn't difficult.

You could also consider the time-honoured lifetime-erasure trick:

`Box::leak(Box::new(<lifetime_bound_object_you_want_to_erase>))`

This will erase an <'a> lifetime and turn it into <'static> instead. But, again, you are deliberately hiding the intent of what you're doing here. There are times when there's no choice, but I just don't think this is one of those times.

**Comment by ajrcarey** [comment]
(Additionally, if you store `Pdfium` in your `Application` struct, and then pass it to `PdfManager` when you create `PdfManager`, you may not need to use a lifetime in `Application` at all, since you aren't storing _references_ to anything.)

**Comment by LU15W1R7H** [comment]
Thanks you very much for your explanations :). I wasn't aware that doing FFI would make the lifetimes so much harder. It's like you said it: The problem lies in my application, and pdfium needing to share memory. I now have a better understanding of the issues at hand.

Thanks for the proposed solutions. I agree that these would be ways forward. It's not true that this crate is unusable for me. I've changed my mind.

I found a different solution now. The crate [`ouroboros`](https://docs.rs/ouroboros/latest/ouroboros/attr.self_referencing.html) allows me to have self-referential structs. My `PdfManager` now looks like this.
```rust
#[ouroboros::self_referencing]
pub struct PdfManager {
  pdfium: Pdfium,

  #[borrows(pdfium)]
  #[covariant]
  document: PdfDocument<'this>,

  #[borrows(document)]
  #[covariant]
  pages: PdfPages<'this>,

  #[borrows(pages)]
  #[covariant]
  page_vec: Vec<PdfPage<'this>>,
}

impl PdfManager {
  pub fn load_document(path: impl AsRef<Path>) -> Self {
    let bindings = Pdfium::bind_to_library("/home/luis/dl/pdfium/lib/libpdfium.so").unwrap();
    let pdfium = Pdfium::new(bindings);

    PdfManagerBuilder {
      pdfium,
      document_builder: |pdfium| pdfium.load_pdf_from_file(&path, None).unwrap(),
      pages_builder: |document| document.pages(),
      page_vec_builder: |pages| pages.iter().collect(),
    }
    .build()
  }

  pub fn page_slice(&self) -> &[PdfPage] {
    self.borrow_page_vec()
  }
}
```

This is now working.
Thanks for your help!

**Comment by LU15W1R7H** [comment]
Here a screenshot of what's now working! 🎉 
![2022-09-15T201538](https://user-images.githubusercontent.com/37505890/190479533-3ffb1d34-800a-468a-bcf0-cbbb0f825c9e.png)

**Comment by ajrcarey** [comment]
Good, I'm really pleased that you have a solution. I am vaguely familiar with `ouroboros`, but haven't used it myself.

One additional comment I'd make about all the lifetime stuff is that it can be hard to see the benefit when you are using immutable objects. If I understand your use case correctly, you are just rendering existing pages, yes? The benefits of lifetimes are more obvious when you are also trying to mutate things as well - the lifetimes really save you from blowing up your application once documents, pages, and objects on those pages are all changing.

I struggled with the issue of lifetimes a lot myself when first implementing `pdfium-render`. It took me a long time to really be convinced that the lifetimes were essential, but once I understood the fundamental problem - memory ownership has to be shared between Pdfium and Rust, how can we do that safely? - I realised just how essential the lifetimes are. Yes, it makes our code more complex, because now we have lifetimes everywhere, and initially it's downright annoying! But every one of those lifetimes is enforcing that memory safety contract between Pdfium and your code. When I think about doing that manually myself, especially when working with mutating documents and pages, I suddenly realise that specifying lifetimes is a really small price to pay for the benefits we gain.

If you have performance issues with `pdfium-render` that I can help you with, or want to revisit the second point you originally raised, about creating a `PdfDocument` from an owned byte buffer rather than a byte slice, please let me know.

If you have a demo of your project available publically, I'd love to see it.

**Comment by LU15W1R7H** [comment]
Yes I am rendering existing static pages for now. But since this is a drawing program, I also want to support exporting the pdf page with the new drawings. But that's for a later time.
Yes I totally see that the lifetimes are correct, safe and sound. It's good that they are enforced. But maybe there is a way to wrap the underlying API some more to get avoid some annoyance when trying to write rust code that uses pdfium. I don't have any ideas how to do this right now, but it might be feasible somehow.

Performance is okay right now, but I will get back to you, if I face any problems. Thanks!

I'm working on getting wasm support for my application. But it's not working too great as of right now. I'll let you know once I have it running.

**Comment by ajrcarey** [comment]
Merged PR #45. Renamed newly public functions to more idiomatic `PdfPage::bindings()`, `PdfPage::document()`. Introduced equivalent `bindings()` public accessor function in `PdfDocument`, `PdfBitmap`, `PdfForm`, `PdfFont`, `PdfPageAnnotations`, `PdfPageBoundaries`, and `PdfPageText` for consistency. Deprecated `Pdfium::get_bindings()` in favour of `Pdfium::bindings()` as part of #36. Updated documentation. Bumped crate version to 0.7.18.

**Comment by ajrcarey** [comment]
Published to crates.io as version 0.7.18.
